### PR TITLE
11 automate spatial feature specification

### PR DESCRIPTION
**Pull request by dimalvovs**
refactor:
- eliminate `featureNames=NULL`  since checking with regex `"."` does the same and is shorter
- simplify and deduplicate `if  else` and messages
- add tests to ensure most cases are covered (Seurat is not well defined so it's not covered)
- introduce helper getters that each work with own input type
- add test objects for cogaps and bayestme

ready to automate spatial feature specification now.

### CODE REVIEW

File: man/cogaps_result.Rd
Comment (atuldeshpande) [inline man/cogaps_result.Rd:1]: Need to delete the corresponding cogaps_result.rda from the data folder? See build check errors.

File: R/preprocessing.R [hunk 1]
```diff
@@ -158,52 +157,79 @@ load10XCoords <- function(visiumDir, resolution = "lowres", version = NULL){
...
+    if (length(featureNames) == 1) {
+        dataNames <- colnames(spFeatures)
+        featureNames <- dataNames[grepl(pattern = namePattern,
+                                        dataNames, ignore.case = TRUE)]
+        message(sprintf("%d features match expression '%s'.",
+                        length(featureNames), namePattern))
     }
-    spFeatures <- spFeatures[,featureNames]
+
+    featureNames <- intersect(featureNames,colnames(spFeatures))
```
Comment (atuldeshpande) [inline R/preprocessing.R:186]: There could be a case where user lists 6 patterns but we find only 5 in the object. How should we handle that?
1. Message which reports that we did not find all patterns, and are continuing with the rest.
2. Error which reports the specific patterns we did not find, and ask user to check input.
I note that this scenario may be more common in non-pipeline use cases.

**Review by atuldeshpande — CHANGES_REQUESTED** [review]
Thank you for this refactoring! It is much easier to follow the logic for each method option even if some method may just be two lines of code.

Please see some comments and suggestions.

File: R/preprocessing.R [hunk 2]
```diff
@@ -158,52 +157,79 @@ load10XCoords <- function(visiumDir, resolution = "lowres", version = NULL){
...
-            spFeatures <- spFeatures[,featureNames]
-        else
-            stop("No features found in the spatial
-                    data with provided feature names.")
+getSpatialFeatures <- function(filePath,method = "CoGAPS",featureNames = "."){
+
+    if(method == "CoGAPS") {
+        spFeatures <- getCogaps(filePath)
+    } else if(method == "BayesTME") {
+        spFeatures <- getBtme(filePath)
```
[the remark this answers is not in the document]
Comment (dimalvovs) [reply R/preprocessing.R:165]: sure, shall the method also become "Anndata"?

File: R/preprocessing.R (hunk 1 above)
Comment (dimalvovs) [reply R/preprocessing.R:186]: Makes sense!

File: man/cogaps_result.Rd
Comment (dimalvovs) [reply man/cogaps_result.Rd:1]: Sure, I did not delete as there is currently a non-merged pull request that does just that. Merging that will resolve the build check error.

File: R/preprocessing.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (atuldeshpande) [reply R/preprocessing.R:165]: I think when we actually add generic Anndata support, we can change method == "BayesTME" to method in c("BayesTME", "Anndata")?

File: man/cogaps_result.Rd
Comment (dimalvovs) [reply man/cogaps_result.Rd:1]: It did not help. I wonder how build-check has worked on the main branch where there is cogaps_result.rda present and no documentation for it.

File: R/preprocessing.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (dimalvovs) [reply R/preprocessing.R:167]: agree, let's change when we actually add generic Anndata support!

File: R/preprocessing.R (hunk 1 above)
Comment (dimalvovs) [reply R/preprocessing.R]: addressed in [cd2ab20](https://github.com/DeshpandeLab/SpaceMarkers/pull/14/commits/cd2ab2089f614a6855336d2f91f07be6515f5b07)

File: man/cogaps_result.Rd
Comment (dimalvovs) [reply man/cogaps_result.Rd:1]: deleted in [b836616](https://github.com/DeshpandeLab/SpaceMarkers/pull/14/commits/b8366163488c6669d57c814286c1b27bdd4d33b4)

**Review by atuldeshpande — APPROVED** [review]

**Pull request merged**